File: plugins/quota-glance/internal/redeem/redeem.go
// Package redeem spends one banked Codex rate-limit reset.
//
// It is the only part of quota-glance that contacts a provider, and it runs
// only when the operator presses the button and confirms. Nothing on a timer,
// nothing on a rebuild, and nothing on a route that merely reads reaches this
// code. quota-cache still owns every scheduled request; this is a write, made
// once, on a person's instruction.
//
// The action is irreversible. A banked reset is consumed the moment the
// provider accepts the POST — there is no unspend — so every guard here is
// about not making that request by accident: the credential must be a Codex
// credential, it must hold a credit that is actually available, and the caller
// must have confirmed. The confirmation itself lives in the browser; what this
// package guarantees is that a request which arrives without one is refused.
package redeem

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"io"
	"strings"
	"sync"
	"time"

	"github.com/NoorChasib/cpa-plugins/plugins/quota-glance/internal/protocol"
)

const (
	creditsURL = "https://chatgpt.com/backend-api/wham/rate-limit-reset-credits"
	consumeURL = creditsURL + "/consume"

	// userAgent is CLI-shaped because the ChatGPT backend edge rejects clients
	// that do not look like one, exactly as quota-cache found on the usage
	// endpoint. The comment segment identifies the real caller.
	userAgent = "codex_cli_rs/0.0.0 (cpa-plugins/quota-glance)"

	// maxResponseBytes bounds a decoded provider response; real ones are a few
	// kilobytes.
	maxResponseBytes = 1 << 20

	// Timeout bounds the whole two-request exchange. It is generous because a
	// caller who has already confirmed would rather wait than be told to press
	// the button again — and pressing it again is the one retry that can cost a
	// second credit.
	Timeout = 30 * time.Second
)

// Outcomes reported to the caller. These are this plugin's vocabulary, not the
// provider's: the provider's own codes are mapped onto them so a new code
// upstream surfaces as a plain failure rather than as an unrecognized success.
const (
	// OutcomeReset is the only outcome that spent a credit.
	OutcomeReset = "reset"
	// OutcomeNothingToReset means the provider accepted the request but no
	// window needed clearing. The credit is still gone: the provider consumes
	// it on acceptance, and saying otherwise would be a lie the operator acts
	// on.
	OutcomeNothingToReset = "nothingToReset"
	// OutcomeNoCredit means there was nothing to spend by the time the request
	// landed — usually a count read from a snapshot that has since been spent
	// elsewhere.
	OutcomeNoCredit = "noCredit"
	// OutcomeFailed is every other ending, including a refusal by the provider
	// and a response this code does not understand.
	OutcomeFailed = "failed"
)

// Errors a caller turns into a status code. Every one of them is a fixed
// string: provider error text is never forwarded, because it is unbounded input
// that would land in a dashboard and a log.
var (
	// ErrNotCodex covers a credential on any other provider. Banked resets are
	// a Codex concept and there is nothing to spend anywhere else.
	ErrNotCodex = errors.New("banked resets exist only on Codex credentials")
	// ErrNoAccessToken and ErrNoAccountID mean the credential cannot
	// authenticate one request — an API-key login has no reset credits at all.
	ErrNoAccessToken = errors.New("credential has no usable Codex access token")
	ErrNoAccountID   = errors.New("credential has no resolvable ChatGPT account id")
	ErrUnavailable   = errors.New("the provider could not be reached")
	ErrRefused       = errors.New("the provider refused the request")
	// ErrInFlight means a redemption against this credential is already under
	// way. Two presses that both reach the provider can spend two credits, and
	// the second one was never intended: a double-click, an impatient retry, a
	// second tab. Refusing the second is the only guard that holds, because the
	// first may already have passed its inventory check.
	ErrInFlight = errors.New("a redemption is already under way for this credential")
)

// Host is the narrow pair of callbacks this package needs. It is an interface
// so the plugin can withhold both when redemption is switched off: a nil Host
// is a redeemer that cannot make a request at all, which is a stronger
// guarantee than a flag consulted at the top of a function.
type Host interface {
	GetAuth(context.Context, string) ([]byte, error)
	HTTPDo(context.Context, protocol.HostHTTPRequest) (protocol.HostHTTPResponse, error)
}

// Result is one redemption attempt that reached the provider.
type Result struct {
	Outcome string
	// WindowsReset is what the provider says it cleared. Zero is normal on
	// OutcomeNothingToReset and is not by itself a failure.
	WindowsReset int
	// RemainingCount is how many credits the account still holds, counted from
	// the inventory read at the start of this exchange minus the one just
	// spent. It is a floor rather than a fresh reading: it costs no third
	// request, and the next quota-cache poll replaces it with the truth.
	RemainingCount int
}

// Redeemer spends credits for one CPA instance.
type Redeemer struct {
	host Host

	// inFlight holds the credentials with a redemption under way, so a second
	// press cannot start a second exchange against the same account while the
	// first is still between its inventory read and its POST.
	mu       sync.Mutex
	inFlight map[string]bool
}

func New(host Host) *Redeemer { return &Redeemer{host: host, inFlight: map[string]bool{}} }

// claim reserves the credential, or reports that it is already reserved.
func (r *Redeemer) claim(authIndex string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.inFlight[authIndex] {
		return false
	}
	if r.inFlight == nil {
		r.inFlight = map[string]bool{}
	}
	r.inFlight[authIndex] = true
	return true
}

func (r *Redeemer) release(authIndex string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.inFlight, authIndex)
}

// Enabled reports whether a redemption could be attempted at all.
func (r *Redeemer) Enabled() bool { return r != nil && r.host != nil }

// Redeem spends one credit against the named credential.
//
// The order is deliberate. The inventory is read first so the credit being
// spent is one the provider currently calls available, rather than one a
// snapshot claimed minutes ago: a POST built from stale evidence is how a
// second credit gets spent on an account that had one. If nothing is available,
// no POST is made at all.
func (r *Redeemer) Redeem(ctx context.Context, provider, authIndex string) (Result, error) {
	if !r.Enabled() {
		return Result{}, ErrUnavailable
	}
	if strings.ToLower(strings.TrimSpace(provider)) != "codex" {
		return Result{}, ErrNotCodex
	}
	if !r.claim(authIndex) {
		return Result{}, ErrInFlight
	}
	defer r.release(authIndex)
	creds, err := credentialsOf(ctx, r.host, authIndex)
	if err != nil {
		return Result{}, err
	}

	available, err := r.available(ctx, creds)
	if err != nil {
		return Result{}, err
	}
	if len(available) == 0 {
		return Result{Outcome: OutcomeNoCredit}, nil
	}

	// A fresh idempotency key per attempt, as the provider's own clients send.
	// It is generated here and never reused: replaying one is the provider's
	// business, and reusing one across two genuine presses would silently drop
	// the second.
	requestID, err := requestID()
	if err != nil {
		return Result{}, ErrUnavailable
	}
	body, err := json.Marshal(map[string]string{"credit_id": available[0], "redeem_request_id": requestID})
	if err != nil {
		return Result{}, ErrUnavailable
	}
	response, err := r.host.HTTPDo(ctx, request("POST", consumeURL, creds, body))
	if err != nil {
		return Result{}, ErrUnavailable
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return Result{}, ErrRefused
	}
	root, err := decode(response.Body)
	if err != nil {
		// The provider accepted it. Whatever the body said, the credit is
		// gone — reporting this as a failure would invite a second press.
		return Result{Outcome: OutcomeReset, RemainingCount: len(available) - 1}, nil
	}
	return Result{
		Outcome:        outcomeOf(stringOf(root, "code")),
		WindowsReset:   intOf(root, "windows_reset", "windowsReset"),
		RemainingCount: len(available) - 1,
	}, nil
}

// outcomeOf maps the provider's code onto this package's vocabulary.
//
// An unrecognized code is treated as a successful reset rather than as a
// failure, because control only reaches here on a 2xx: the credit has been
// spent, and the only question left is what to print. Calling that a failure
// would be the one mistake that costs a second credit.
func outcomeOf(code string) string {
	switch code {
	case "reset":
		return OutcomeReset
	case "nothing_to_reset", "nothingToReset":
		return OutcomeNothingToReset
	case "no_credit", "noCredit", "already_redeemed", "alreadyRedeemed":
		return OutcomeNoCredit
	default:
		return OutcomeReset
	}
}

// available lists the ids of credits the provider currently calls spendable,
// soonest expiry first, so the credit closest to lapsing is the one spent.
func (r *Redeemer) available(ctx context.Context, creds credentials) ([]string, error) {
	response, err := r.host.HTTPDo(ctx, request("GET", creditsURL, creds, nil))
	if err != nil {
		return nil, ErrUnavailable
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return nil, ErrRefused
	}
	root, err := decode(response.Body)
	if err != nil {
		return nil, ErrRefused
	}
	list, _ := root["credits"].([]any)
	type candidate struct {
		id      string
		expires time.Time
	}
	candidates := make([]candidate, 0, len(list))
	for _, item := range list {
		credit, ok := item.(map[string]any)
		if !ok || stringOf(credit, "status") != "available" {
			continue
		}
		id := stringOf(credit, "id")
		// Bounded, and matched against the shape the provider actually sends.
		// An id is echoed straight back in a request body; an unbounded string
		// from a response has no business being one.
		if id == "" || len(id) > 128 || strings.ContainsAny(id, "\x00\n\r\"\\") {
			continue
		}
		at, _ := time.Parse(time.RFC3339, stringOf(credit, "expires_at"))
		candidates = append(candidates, candidate{id: id, expires: at})
	}
	// Undated credits sort last: a known deadline is a reason to spend one
	// first, and an unknown one is not a reason to spend it before a dated one.
	sortStable(candidates, func(a, b candidate) bool {
		if a.expires.IsZero() != b.expires.IsZero() {
			return !a.expires.IsZero()
		}
		return a.expires.Before(b.expires)
	})
	ids := make([]string, 0, len(candidates))
	for _, one := range candidates {
		ids = append(ids, one.id)
	}
	return ids, nil
}

// credentials is the pair of values one Codex request needs. Nothing else is
// kept from the credential document, and this value never leaves the package.
type credentials struct {
	accessToken string
	accountID   string
}

// credentialsOf decodes only the two fields a request needs out of the physical
// credential document. The document holds OAuth tokens in full and is never
// logged, persisted, rendered, or returned.
func credentialsOf(ctx context.Context, host Host, authIndex string) (credentials, error) {
	raw, err := host.GetAuth(ctx, authIndex)
	if err != nil {
		return credentials{}, ErrUnavailable
	}
	var doc struct {
		AccessToken string `json:"access_token"`
		AccountID   string `json:"account_id"`
		IDToken     string `json:"id_token"`
		// Codex CLI nests the same fields one level down.
		Tokens *struct {
			AccessToken string `json:"access_token"`
			AccountID   string `json:"account_id"`
			IDToken     string `json:"id_token"`
		} `json:"tokens"`
	}
	if json.Unmarshal(raw, &doc) != nil {
		return credentials{}, ErrNoAccessToken
	}
	out := credentials{accessToken: doc.AccessToken, accountID: doc.AccountID}
	idToken := doc.IDToken
	if doc.Tokens != nil {
		if out.accessToken == "" {
			out.accessToken = doc.Tokens.AccessToken
		}
		if out.accountID == "" {
			out.accountID = doc.Tokens.AccountID
		}
		if idToken == "" {
			idToken = doc.Tokens.IDToken
		}
	}
	if out.accessToken == "" {
		return credentials{}, ErrNoAccessToken
	}
	if out.accountID == "" {
		out.accountID = accountIDFromIDToken(idToken)
	}
	if out.accountID == "" {
		return credentials{}, ErrNoAccountID
	}
	return out, nil
}

func request(method, url string, creds credentials, body []byte) protocol.HostHTTPRequest {
	headers := map[string][]string{
		"Authorization":      {"Bearer " + creds.accessToken},
		"Accept":             {"application/json"},
		"User-Agent":         {userAgent},
		"Chatgpt-Account-Id": {creds.accountID},
	}
	if body != nil {
		headers["Content-Type"] = []string{"application/json"}
	}
	return protocol.HostHTTPRequest{Method: method, URL: url, Headers: headers, Body: body}
}

func decode(body []byte) (map[string]any, error) {
	if len(body) > maxResponseBytes {
		return nil, errors.New("response too large")
	}
	decoder := json.NewDecoder(bytes.NewReader(body))
	decoder.UseNumber()
	var root map[string]any
	if err := decoder.Decode(&root); err != nil || root == nil {
		return nil, errors.New("unreadable response")
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		return nil, errors.New("unreadable response")
	}
	return root, nil
}

func stringOf(obj map[string]any, keys ...string) string {
	for _, key := range keys {
		if value, ok := obj[key].(string); ok {
			return value
		}
	}
	return ""
}

func intOf(obj map[string]any, keys ...string) int {
	for _, key := range keys {
		if value, ok := obj[key].(json.Number); ok {
			if n, err := value.Int64(); err == nil && n >= 0 && n < 1<<20 {
				return int(n)
			}
		}
	}
	return 0
}

// requestID is a random 128-bit value in the hyphenated form the provider's own
// clients send. It identifies one attempt and carries nothing about the caller.
func requestID() (string, error) {
	raw := make([]byte, 16)
	if _, err := rand.Read(raw); err != nil {
		return "", err
	}
	// Version 4, variant 1, so a server that validates the shape accepts it.
	raw[6] = raw[6]&0x0f | 0x40
	raw[8] = raw[8]&0x3f | 0x80
	hexed := hex.EncodeToString(raw)
	return strings.Join([]string{hexed[:8], hexed[8:12], hexed[12:16], hexed[16:20], hexed[20:]}, "-"), nil
}

// sortStable is an insertion sort over a handful of credits, kept here so the
// comparison reads beside the rule it implements.
func sortStable[T any](items []T, less func(a, b T) bool) {
	for i := 1; i < len(items); i++ {
		for j := i; j > 0 && less(items[j], items[j-1]); j-- {
			items[j], items[j-1] = items[j-1], items[j]
		}
	}
}

File: plugins/quota-glance/web/src/lib/redeem.ts
// Spending one banked Codex rate-limit reset.
//
// This is the only request this page makes that changes anything, and it
// cannot be undone: the credit is gone the moment the provider accepts it. So
// it is never retried, never made without the dialog's answer, and never made
// twice from one press — `confirmed` is carried in the body so a request that
// somehow arrives without one is refused by the plugin rather than assumed.

import { authHeaders, managementKey, managementPath } from "./cpa-auth"
import * as token from "./token"

/** What the plugin reports back. Outcomes widen additively, like every enum. */
export type RedeemOutcome = "reset" | "nothingToReset" | "noCredit" | (string & {})

export interface RedeemResult {
  outcome: RedeemOutcome
  windowsReset: number
  remainingCount: number
  /** The card's count comes from quota-cache and lags until its next poll. */
  snapshotPending: boolean
}

/**
 * A failure with something worth printing.
 *
 * `code` is one of the plugin's own fixed codes — never provider text, which
 * the plugin deliberately does not forward. Anything unrecognized falls through
 * to a generic sentence rather than putting a raw code on screen.
 */
export class RedeemError extends Error {
  constructor(
    readonly code: string,
    readonly status: number,
  ) {
    super(code)
    this.name = "RedeemError"
  }
}

const RESOURCE_FALLBACK = "/v0/resource/plugins/quota-glance/redeem"

function fallbackURL(): string {
  const path = window.location.pathname.replace(/\/+$/, "")
  return path.endsWith("/app") ? `${path.slice(0, -"/app".length)}/redeem` : RESOURCE_FALLBACK
}

/**
 * Where to send it, best first — the same order, and the same reasoning, as
 * reading the document: the console's key costs the reader nothing, and the
 * saved password is the only thing a reader without a console session has.
 *
 * Unlike a read, this is not retried down the list on rejection. A POST that
 * was refused may still have spent the credit, and trying the second door would
 * risk spending a second one to find out.
 */
function target(): { url: string; headers: Record<string, string>; viaConsole: boolean } | null {
  const headers = { "Content-Type": "application/json", Accept: "application/json" }
  if (managementKey() !== null) {
    return { url: managementPath("/redeem"), headers: authHeaders(headers), viaConsole: true }
  }
  const saved = token.read()
  if (saved) {
    return {
      url: fallbackURL(),
      headers: { ...headers, Authorization: ["Bearer", saved].join(" ") },
      viaConsole: false,
    }
  }
  return null
}

/**
 * Whether this browser can redeem at all.
 *
 * CPA dispatches only GET to a resource route — verified against a real CPA by
 * scripts/quota-glance-smoke.py, which reports a 404 on the public redeem path
 * — so the `web-token` door can read the document but cannot spend anything.
 * The count and the expiry still show there; the button does not, because a
 * control that is certain to fail is worse than no control.
 */
export function canRedeemHere(): boolean {
  return managementKey() !== null
}

/**
 * Carries ?redeem= through to the dev fixture route, which is how each ending
 * is reviewed. Stripped from the production bundle, where the only thing on the
 * other end is the plugin.
 */
function devEnding(url: string): string {
  if (!import.meta.env.DEV) return url
  const ending = new URLSearchParams(window.location.search).get("redeem")
  return ending === null ? url : `${url}?redeem=${encodeURIComponent(ending)}`
}

async function codeOf(response: Response): Promise<string> {
  try {
    const body = (await response.json()) as { error?: unknown }
    if (typeof body.error === "string" && body.error !== "") return body.error
  } catch {
    /* no body, or not JSON */
  }
  return "unknown"
}

export async function redeemReset(credentialId: string): Promise<RedeemResult> {
  const one = target()
  if (one === null) throw new RedeemError("no_session", 0)

  let response: Response
  try {
    response = await fetch(devEnding(one.url), {
      method: "POST",
      headers: one.headers,
      // The plugin refuses a body without this, so the dialog's answer travels
      // with the request rather than being implied by it having been sent.
      body: JSON.stringify({ credentialId, confirmed: true }),
      signal: AbortSignal.timeout(35_000),
      cache: "no-store",
      credentials: "same-origin",
    })
  } catch {
    // The request never completed. Whether it landed is genuinely unknown, and
    // the caller says so rather than inviting a second press.
    throw new RedeemError("unreachable", 0)
  }

  if (!response.ok) {
    // A 404 means two different things by door: from the console it is the
    // route being switched off, and from the token door it is CPA never having
    // dispatched the POST. Naming the right one is the difference between the
    // reader changing a setting and the reader signing in.
    if (response.status === 404 && !one.viaConsole) throw new RedeemError("needs_console", 404)
    throw new RedeemError(await codeOf(response), response.status)
  }
  const body = (await response.json()) as Partial<RedeemResult>
  return {
    outcome: typeof body.outcome === "string" ? body.outcome : "reset",
    windowsReset: typeof body.windowsReset === "number" ? body.windowsReset : 0,
    remainingCount: typeof body.remainingCount === "number" ? body.remainingCount : 0,
    snapshotPending: body.snapshotPending !== false,
  }
}

/** What to tell the reader, in whole sentences, for each ending. */
export function describeOutcome(result: RedeemResult): string {
  const left =
    result.remainingCount === 0
      ? "No banked resets left."
      : `${result.remainingCount} banked reset${result.remainingCount === 1 ? "" : "s"} left.`
  const lag = result.snapshotPending ? " The card updates at the next quota-cache poll." : ""
  switch (result.outcome) {
    case "reset":
      return `Reset applied. ${left}${lag}`
    case "nothingToReset":
      // The credit is gone either way. Saying otherwise would be a lie the
      // reader acts on when they check their windows.
      return `The credit was spent, but no window needed clearing. ${left}${lag}`
    case "noCredit":
      return "There was nothing left to spend — the count on this card was out of date."
    default:
      return `Done. ${left}${lag}`
  }
}

/** What to tell the reader when it failed, per the plugin's own codes. */
export function describeError(error: unknown): string {
  if (!(error instanceof RedeemError)) return "Something went wrong. Nothing was spent."
  switch (error.code) {
    case "no_session":
      return "You are signed out. Sign in again before redeeming."
    case "unreachable":
      // Deliberately not "nothing was spent": the request may have landed.
      return "The request did not complete, so whether the reset was applied is unknown. Check your Codex usage before trying again."
    case "provider_refused":
      return "Codex refused the request. Nothing was spent."
    case "provider_unavailable":
      return "Codex could not be reached. Nothing was spent."
    case "already_in_flight":
      return "A redemption is already running for this credential."
    case "not_redeemable":
      return "This credential has nothing to redeem right now."
    case "credential_unusable":
      return "This credential cannot redeem a reset — an API-key login has none."
    case "confirmation_required":
      return "The confirmation did not reach the plugin. Nothing was spent."
    case "needs_console":
      return "Redeeming needs a CPA console session; this browser is signed in with the fallback password. Nothing was spent."
    default:
      return error.status === 404
        ? "Redeeming is switched off for this dashboard."
        : "The reset could not be applied. Nothing was spent."
  }
}
